src/server/player_sao.cpp

```diff
@@ -558,12 +558,35 @@ void PlayerSAO::setMaxSpeedOverride(const v3f &vel)
 
 bool PlayerSAO::checkMovementCheat()
 {
-	if (isAttached() || m_is_singleplayer ||
+	if (m_is_singleplayer ||
 			g_settings->getBool("disable_anticheat")) {
 		m_last_good_position = m_base_position;
 		return false;
 	}
 
+	ServerActiveObject *parentSAO = getParent();
+
+	if (parentSAO) {
+		UnitSAO *parent = dynamic_cast<UnitSAO *>(parentSAO);
+		if (! parent)
+			return false;
+		int parent_id;
+		std::string bone;
+		v3f attachment_pos, attachment_rot;
+		getAttachment(&parent_id, &bone, &attachment_pos, &attachment_rot);
+		v3f parent_pos = parent->getBasePosition();
+		f32 diff = (getBasePosition() - parent_pos).getLength() - (attachment_pos * 3).getLength();
+		if (diff > 10) {
```

Comment:
This
* doesn't handle negative numbers
* `attachment_pos * 3` already leaves some room for uncertainty, why have more here?

Reply:
I think both of the values still need tuning:

With `attachment_pos * 3` changed to `attachment_pos * 1`, `diff > 2 * BS` is enough to avoid false-positives in singleplayer. Haven't tested with latency yet.

Reply:
it can be interesting to see what happen when you are in a cart which accelerate massively due to the path

Reply:
Ok, after testing in singleplayer with different latencies this is my suggestion:
```diff
+	if (parentSAO) {
+		UnitSAO *parent = dynamic_cast<UnitSAO *>(parentSAO);
+		if (!parent)
+			return false;
+		int parent_id;
+		std::string bone;
+		v3f attachment_pos, attachment_rot;
+		getAttachment(&parent_id, &bone, &attachment_pos, &attachment_rot);
+
+		v3f parent_pos = parent->getBasePosition();
+		f32 diff = m_base_position.getDistanceFromSQ(parent_pos) - attachment_pos.getLengthSQ();
+		const f32 maxdiff = 4 * BS;
+		if (diff > maxdiff * maxdiff) {
+			setBasePosition(parent_pos);
+			actionstream << "Server: " << m_player->getName()
+					<< " moved away from parent; diff=" << sqrtf(diff) / BS
+					<< " resetting position." << std::endl;
+			return true;
+		}
+		return false;
+	}
```